### DISCUSSION TITLE
Testing local changes

### DISCUSSION THREAD

**Discussion by antoniolinhart** [google/mdbook-i18n-helpers#73] [Q&A] [answered]
Hi all!

Do you have any tips for testing the executables in this helpers repo? I am working on #67 and currently have some changes made in the main function to `xgettext.rs` to expect another parameter in the config from stdin.

It looks like it may be possible to call it directly, I just cannot find the correct shape for the RenderContext config:
```
Caused by:
    0: Unable to deserialize the `RenderContext`
    1: missing field `root` at line 1 column 75
```

Is there an example 'input' to this executable that someone could give me the template for? Or alternatively, is there a way that I can use this 'local' version of mdbook-i18n-helpers in a repository that uses it already like [Comprehensive Rust](https://github.com/google/comprehensive-rust)? I was looking at this, but since Comprehensive Rust uses a specific version I wasn't sure if I could point the dependency on mdbook-i18n-helpers to a specific commit or similar in this repository.

Thanks for the help!
Antonio

**Comment by antoniolinhart** [comment, marked as answer]
Figured out the solution :smile: 

For others who are interested:
* Clone the `comprehensive-rust` repository (or the repository that makes use of this that you want to use)
* Override the version of `mdbook-i18n-helpers` with the local copy (keep in mind you will need to `cargo install mdbook-i18n-helpers` when you want to go back to the real version
    * `cargo install --path ../mdbook-i18n-helpers/`
* .. for `xgettext` specifically, set the config and then build the book
    * `MDBOOK_OUTPUT='{"xgettext": {"pot-file": "messages.pot"}}'   mdbook build -d po`

**Reply by mgeisler** [reply]
Hi @antoniolinhart!

Sorry for not coming back to this earlier — but great that you found the solution. Basically, when you run `mdbook`, it will use the `mdbook-xgettext` binary from `$PATH`. So you can switch between different versions by running `cargo install --path .` inside the `mdbook-i18n-helpers` directory as you found out.

The above works for testing a binary. There is a related trick to testing fixes to Rust _libraries_. These are normally specified in the `Cargo.toml` file as

```toml
[dependencies]
foo = "1.2.3"
```

You can change this to a [path dependency](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-path-dependencies) and make the program use a customized library from your machine:

```toml
[dependencies]
foo = { path = "../foo" }
```

**Comment by mgeisler** [comment]
> I am working on #67 and currently have some changes made in the main function to `xgettext.rs` to expect another parameter in the config from stdin.

As a side-note, the way to test this would be to write a unit test. The `mdbook-xgettext` binary already has a few tests, but they test the functionality of the code _after_ creating a `RenderContext`. You can also see how the binary is reading config values from `ctx` in its `main` function. I think you'll want to read more configuration values in the same manner:

```rust
    let split_depth = cfg
        .get("split-depth")
        .ok_or_else(|| anyhow!("Missing output.xgettext.split-depth config value"))?
        .as_int() // or similar, I didn't look up the toml API
        .ok_or_else(|| anyhow!("Expected an integer for output.xgettext.split-depth"))?;
```

You should probably not error out if the value isn't set: instead use a default so that users get the current behavior.
